lib/bridge/IrcHandler.js

```diff
@@ -201,6 +207,55 @@ IrcHandler.prototype.onMessage = Promise.coroutine(function*(req, server, fromUs
         return;
     }
 
+    // For topics, set matrix topic as AS bot user
+    if (action.type === "topic") {
+        // Only bridge topics for rooms created by the bridge, via !join or an alias
+        let origins = ["join", "alias"];
+        let mappings = yield this.ircBridge.getStore().getMappingsForChannelByOrigin(
+            server, channel, origins
+        );
+        let promises = mappings.map((entry) => {
+
+            if (this._roomIdToTopic[entry.matrix.getId()] === action.text) {
+                req.log.info(
+                    "Topic of %s already set to '%s'",
+                    entry.matrix.getId(),
+                    action.text
+                );
+                return;
+            }
+
+            req.log.info(
+                "Message is a topic, using AS bot to set topic in room %s: %s",
+                entry.matrix.getId(),
+                action.text
+            );
+
+            return this.ircBridge.getAppServiceBridge().getIntent().setRoomTopic(
```

Comment:
I tried to warn you about this yesterday, but maybe I wasn't clear enough. This code will race.

Even *with* the guard at `this._roomIdToTopic`, you can still set the **same** room topic multiple times. Do you see why? I've highlighted this line for a reason :)